src/main/java/com/iexec/core/tools/ContextualLock.java

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright 2022 IEXEC BLOCKCHAIN TECH
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.iexec.core.tools;
+
+import com.iexec.common.utils.ContextualLockRunner;
+import lombok.extern.slf4j.Slf4j;
+import net.jodah.expiringmap.ExpiringMap;
+
+import java.time.Duration;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+
+@Slf4j
+public class ContextualLock<K> {
+    private final ContextualLockRunner<K> contextualLockRunner =
+            new ContextualLockRunner<>(1, TimeUnit.MINUTES);
+
+    private final Map<K, Boolean> locks;
+
+    public ContextualLock(Duration lockDuration) {
+        locks = ExpiringMap.builder()
+                .expiration(lockDuration.getSeconds(), TimeUnit.SECONDS)
+                .build();
+    }
+
+    /**
+     * Gets a lock on a particular key if no lock currently exists on that key.
+     * <p>
+     * This method is synchronized on the key,
+     * so we're sure there shouldn't be any race condition.
+     *
+     * @param key Key to lock on.
+     * @return {@literal true} if the lock has been acquired,
+     * {@literal false} otherwise.
+     */
+    public boolean lockIfPossible(K key) {
```

Comment:
What about adding at least one UT with a 1st locking thread + a 2nd one trying to lock (but fail) without waiting. E.g I see a thread2 which loops by doing lockIfPossible which always fail but only succeeds right after thread1 unlocks.

Reply:
Good idea. See https://github.com/iExecBlockchainComputing/iexec-core/pull/490/commits/009e861c03e999070b21aa8443d1563ef2e49a4a.